hw/dv/dpi/gpiodpi/gpiodpi.c

```diff
@@ -21,83 +21,238 @@
 #include <sys/types.h>
 #include <unistd.h>
 
+// This file does a lot of bit setting and getting; these macros are intended to
+// make that a little more readable.
+#define GET_BIT(word, bit_idx) ((word) >> bit_idx) & 1
+#define SET_BIT(word, bit_idx) word |= (1 << bit_idx);
+#define CLR_BIT(word, bit_idx) word &= ~(1 << bit_idx);
+
+typedef struct gpiodpi_ctx {
+  // The number of pins we're driving.
+  int n_bits;
+
+  // The last known value of the pins, in little-endian order.
+  uint32_t driven_pin_values;
+
+  // File descriptors and paths for the device-to-host and host-to-device
+  // FIFOs.
+  int dev_to_host_fifo;
+  char dev_to_host_path[PATH_MAX];
+  int host_to_dev_fifo;
+  char host_to_dev_path[PATH_MAX];
+} gpiodpi_ctx_t;
```

Comment:
I'm not a fan of typedef'ing structs, and not a fan of naming them `_t` either (which is reserved by POSIX). But leaving that up to you for final decision.

Reply:
This is used pretty consistently in our other C code, so I don't think it's an issue.

Reply:
Not really, all other (context) structs in the DPI modules don't use typedefs. But fine either way.

Reply:
Meh. The DPI files are so far out-of-style that I think I'll hold off on the typedef until a future cleanup of all this stuff.